src/regex.nim

```diff
@@ -2100,6 +2101,79 @@ proc groupsCount*(m: RegexMatch): int =
   ##
   m.groups.len
 
+
+proc groupNames*(m: RegexMatch): seq[string] =
+  ## return the names of capturing groups
+  ##
+  ## .. code-block:: nim
+  ##   block:
+  ##     let text = "hello world her"
+  ##     var m: RegexMatch
+  ##     doAssert text.match(re"(?P<greet>hello) (?P<who>world) (?P<who>her)", m)
+  ##     for group in @["greet", "who"]:
+  ##       doAssert m.groupNames.contains(group)
+  result = toSeq(m.namedGroups.keys)
+
+proc group*(m: RegexMatch, groupName: string, text:string): seq[string] = 
+  ## return seq of captured text by group `groupName`
+  ##
+  ## .. code-block:: nim
+  ##   block:
+  ##     let text = "hello world"
+  ##     var m: RegexMatch
+  ##     doAssert text.match(re"(?P<greet>hello) (?P<who>world)", m)
+  ##     doAssert m.group("greet", text) == @["hello"]
+  ##     doAssert m.group("who", text) == @["world"]
+  ##   block:
+  ##     let text = "hello world foo bar"
+  ##     var m: RegexMatch
+  ##     doAssert text.match(re"(?P<greet>hello) (?:(?P<who>[^\s]+)\s?)+", m)
+  ##     doAssert m.group("greet", text) == @["hello"]
+  ##     let whoGroups = m.group("who", text)
+  ##     for w in @["foo", "bar", "world"]:
+  ##       doAssert whoGroups.contains(w)
+  result = newSeq[string]()
+  for bounds in m.group(groupName):
+    result.add text[bounds]
+
+proc groupFirstCapture*(m: RegexMatch, groupName: string, text: string): string =
+  ##  Return last capture for a given capturing group
+  ##
+  ## .. code-block:: nim
+  ##   block:
+  ##     let text = "hello world her"
+  ##     var m: RegexMatch
+  ##     doAssert text.match(re"(?P<greet>hello) (?P<who>world) (?P<who>her)", m)
+  ##     doAssert m.groupFirstCapture("greet", text) == "hello"
+  ##   block:
+  ##     let text = "hello world foo bar"
+  ##     var m: RegexMatch
+  ##     doAssert text.match(re"(?P<greet>hello) (?:(?P<who>[^\s]+)\s?)+", m)
+  ##     # "who" captures @["world", "foo", "bar"]
+  ##     doAssert m.groupFirstCapture("who", text) == "world"
+  
+  m.group(groupName, text)[0]
```

Comment:
A group may not contain any captures (i.e: when using the `?` operator), then `group` will return an empty seq. We need to check for this, and return an empty string in that case. Same for `groupLastCapture`.

Test case:

```nim
let text = "hello"
var m: RegexMatch
doAssert text.match(re"(?P<greet>hello)\s?(?P<who>world)?", m)
doAssert m.groupFirstCapture("greet", text) == "hello"
doAssert m.groupFirstCapture("who", text) == ""
```

Reply:
updated to return empty string